### ISSUE TITLE
About the Affine layer

### ISSUE THREAD

**Issue by coocoo90** [facebookresearch/video-long-term-feature-banks#31]
Hi, sorry to bother you.
I'm trying to transfer the model to pytorch version but I'm not sure whether I understood the Affine layer correctly. Is the Affine layer a batch normalization layer that freezes the gamma and beta value and without tracking runnable status?

**Comment by chaoyuaw** [comment]
Hi @coocoo90 , yes, Affine layer is a frozen batch norm. The gamma/beta and running statistics are all fixed, so it reduces an affine transform.

The combination of gamma/beta and running mean/var is done at 
https://github.com/facebookresearch/video-long-term-feature-banks/blob/master/lib/utils/checkpoints.py#L88

**Comment by coocoo90** [comment]
Thanks for the answer. I have reimplemented the affine layer based on your instruction.
But got another question about the weights with "momentum" postfix such as "nonlocal_conv3_1_out_w_momentum".

I thought these weights only needs to be removed but I got confused when I trying to load the weights from nonlocal blocks. It seems some layer only have weights with "momentum". 
For example, in the baseline model "charades_r101_baseline", at nonlocal block "nonlocal_conv3_1", there are only weights for "nonlocal_conv3_1_g_w', "nonlocal_conv3_1_out" and "nonlocal_conv3_1_phi" without "nonlocal_conv3_1_theta". 
But there are "nonlocal_conv3_1_theta_w_momentum" and "nonlocal_conv3_1_theta_b_momentum". 

Is there any special merging method for merge weights from regular and "momentum'?
Sorry for bothering you again. This is an excellent work, I really want to transfer it into PyTorch.

**Comment by chaoyuaw** [comment]
Hi @coocoo90 , many thanks for your efforts porting this to PyTorch! :) 

I'm not sure if I understand your question correctly:
When I do
`[x for x in data['blobs'] if 'nonlocal_conv3_1' in x]`
I see
`
nonlocal_conv3_1_bn_b

nonlocal_conv3_1_bn_s

nonlocal_conv3_1_g_b

nonlocal_conv3_1_g_b_momentum

nonlocal_conv3_1_g_w

nonlocal_conv3_1_g_w_momentum

nonlocal_conv3_1_out_b

nonlocal_conv3_1_out_b_momentum

nonlocal_conv3_1_out_w

nonlocal_conv3_1_out_w_momentum

nonlocal_conv3_1_phi_b

nonlocal_conv3_1_phi_b_momentum

nonlocal_conv3_1_phi_w

nonlocal_conv3_1_phi_w_momentum

nonlocal_conv3_1_theta_b

nonlocal_conv3_1_theta_b_momentum

nonlocal_conv3_1_theta_w

nonlocal_conv3_1_theta_w_momentum
`

Could you please elaborate on your question?

**Comment by coocoo90** [comment]
Hi @chaoyuaw, thanks for the clarifying.
I find out the problem. 
It seems I didn't load the complete model file previously. Because I'm using  python3 instead of python2 (there will be an error if I load the .pkl file using python3 directly), I need to transfer the ".pkl" file to ".mat" using python2 environment then load it using python3. 
I just can't see the "nonlocal_conv3_1_theta_w" and "nonlocal_conv3_1_theta_b" at the previous transferred ".mat" file. That's why I got confused.
I got the complete one after trying it again, although I still don't know why they are lost during the previous transferring Lol.

Just double confirm, all the "***_momentum" can be ignored right?

**Comment by chaoyuaw** [comment]
Hi @coocoo90 , Glad that you figured it out.

If you just want to evaluate the models, yes, the momentums are not needed. However, if you would like to continue training from these model weights, I think it's better to have these momentums.

**Comment by coocoo90** [comment]
Hi, @chaoyuaw again Lol.
I have successfully got the exactly same result after porting the model to the PyTorch version.
I got 38.3% MAP for res-50 nl and 40.38% for res-101 nl, thanks for your instruction. I'm trying to do the same thing for kinetics-400. I have tried using the same config as the charade for evaluating kinetics. But it seems the model only got around 50% accuracy. I think it might be some parameters unmatched such as the "dilation after conv5". 
Is that possible to share your config file when you are pretraining the model on kinetics-400?
I'm not sure whether it needs to be held because of the privacy, it would be perfect if you can provide the configuration file for the k400.

**Comment by chaoyuaw** [comment]
Hi @coocoo90 , Glad to hear that you reproduced the baseline numbers! 

The architecture for K400-pre-training is from the non-local net paper. You can compare the architecture table in non-local paper, and the one in LFB paper's supplementary material/appendix (Table 5 of https://arxiv.org/pdf/1812.05038.pdf).  Dilation is 1 for Kinetics.

**Comment by coocoo90** [comment]
Hi @chaoyuaw thanks for the answer. I will double-check the configurations in the LFB paper again.
And BTW, when you evaluating the kinetics dataset, is that the same with charades for sampling 10 clips and 3-crops per video and combine the predictions using max-pooling? 
And the mean and std for input normalizing are 0.45 and 0.225.

**Comment by chaoyuaw** [comment]
Hi @coocoo90,  Yes, Kinetics uses the same 3-crop * 10 clips testing, but combined by average pooling (following prior work, e.g. non-local). Yes, input normalization mean/std are the same.

**Comment by coocoo90** [comment]
Hi @chaoyuaw Thanks for the confirm. I still didn't find out the issue (〒︿〒).
I have tried to fine-tune the model on Kinetics but only got around 74% on video level.
If it is correct, is the video level accuracy for Res-101 NL same as the one reported on the non-local paper (77.7% top 1 and 93.3 top5) ?

**Comment by chaoyuaw** [comment]
Hi @coocoo90, the pre-trained models should work on K400 without further fine-tuning. Maybe it worths checking https://github.com/Tushar-N/pytorch-resnet3d
I did not read too carefully, but my understanding is that Tushar's repository should work with the pre-trained models we provide here. I have check for the exact number for R101, but the expected R50 accuracy is reported in LFB paper.

**Comment by chaoyuaw** [comment]
Closing it now, but feel free to reopen if you have additional questions.
